### PR TITLE
feat: Enable `FreeParameter` creation for trainable parameters

### PR DESCRIPTION
**Pull request by speller26**
*Description of changes:*

Creates a `FreeParameter` for every parameter marked with `requires_grad=True`; these parameters will be bound service-side.

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-pennylane-plugin-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-pennylane-plugin-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-pennylane-plugin-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-pennylane-plugin-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by kshitijc — COMMENTED** [review]
Could you please add more context on how we expect to use the parametrize_differentiable arg during the execution workflow?

File: src/braket/pennylane_plugin/braket_device.py [hunk 1]
```diff
@@ -87,6 +87,9 @@ class BraketQubitDevice(QubitDevice):
             ``0``, the device runs in analytic mode (calculations will be exact).
         noise_model (NoiseModel or None): The Braket noise model to apply to the circuit before
             execution.
+        parametrize_differentiable (bool): Whether to bing differentiable parameters (parameters
```
[the remark this answers is not in the document]
Comment (speller26) [reply src/braket/pennylane_plugin/braket_device.py:90]:
```suggestion
        parametrize_differentiable (bool): Whether to bind differentiable parameters (parameters
```

File: test/unit_tests/test_braket_device.py [hunk 2]
```diff
@@ -321,6 +321,48 @@ def test_execute(mock_run):
...
+        qml.sample(qml.PauliZ(3))
+
+    dev.execute(circuit)
+
+    assert dev.task == TASK
+    EXPECTED_CIRC = (
+        Circuit()
+        .h(0)
+        .unitary([0], 1 / np.sqrt(2) * np.array([[1, 1], [1, -1]]))
+        .rx(0, FreeParameter("p_1"))
```
Comment (kshitijc) [inline test/unit_tests/test_braket_device.py:346]: Based on the doc string for `parametrize_differentiable` I would have expected this to happen only for parameters marked with `requires_grad=True`, which this wasn't?

Comment (speller26) [reply test/unit_tests/test_braket_device.py:346]: When using `QuantumTape` directly, all parameters are automatically considered differentiable; however, this is not the "normal" way to use PennyLane. If the circuit is constructed as a qnode, then the `qml.RX`  parameter will need to be explicitly marked. I can add a comment for this.

Comment (speller26) [reply test/unit_tests/test_braket_device.py:346]:
```suggestion
        # When using QuantumTape directly (as opposed to a QNode),
        # all parameters are automatically considered differentiable
        .rx(0, FreeParameter("p_1"))
```

File: src/braket/pennylane_plugin/braket_device.py [hunk 3]
```diff
@@ -87,6 +87,9 @@ class BraketQubitDevice(QubitDevice):
             ``0``, the device runs in analytic mode (calculations will be exact).
         noise_model (NoiseModel or None): The Braket noise model to apply to the circuit before
             execution.
+        parametrize_differentiable (bool): Whether to bind differentiable parameters (parameters
+            marked with ``required_grad=True``) on the Braket device rather than in PennyLane.
+            Default: False.
```
Comment (kshitijc) [inline src/braket/pennylane_plugin/braket_device.py:92]: I'm not sure whether the default should be True or False. We should evaluate this based on the typical user workflow during the requires_grad=True execution path, along with performance considerations.

**Review by kshitijc — DISMISSED** [review]

**Comment by yitchen-tim** [comment]
The parameters are only converted to FreeParameter in the execute method. When user do `parallel=True` , the device uses `batch_execute` which does not convert to  FreeParameter .
Action: implement conversion in `batch_execute()`

**Review by kshitijc — APPROVED** [review]

**Comment by kshitijc** [comment]
Could we test these changes with parametric pulse gates?

**Pull request merged**